README.md
## CatFaces - Train a neural network to detect cat fur 
This repo includes the source code to train a neural network to detect my cats name based on their fur. This approach can be applied to any monitoring camera with limited and recurrent traffic detected.

There are 4 main steps : 
 1. Prepare the dataset extracted from a video monitoring file created through motion detection ;
 2. Train & export the deep neural network to be able associate cat fur to a cat name (this requires a GPU)
 3. Test the neural network against a new dataset through command line
 4. Create a microservice able to respond to REST API call to detect a cat in a picture 

### 1. Prepare the dataset
 The dataset is based on a video monitoring file which is built from the Motion software. Each frame is extracted in a folder ('images/'); for example in the picture below #1 and #2 are frames directly extracted from the video file through the following command line : 

> ffmpeg -i cat.avi cat%4d.jpg

Then a reference frame is chosen where there is no motion object : here frame #1 is our 'reference.jpg' file which will be used to detect any motion objects. The masking techniques are applied (images #3 and #4) to detect pixel difference between the 2 pictures and only large deltas are kept. This leads to picture #5 detecting the cat itself; it is then easy to extract the sample as in #6.

Batch processing can be achieved through this command line:
> python process.py -r reference.jpg -s images/ -t extract/

Have a look to this repository to check the extracts thru this process : https://github.com/besn0847/catfaces/tree/master/dataset/mikado

Finally, you have to classify one by one each thumbnail in the dataset folder with the name of the animal. The dataset is ready for training.

![Data extraction process](https://github.com/besn0847/catfaces/raw/master/data_extraction.png)
### 2. Train & export the deep neural network
Next step is to train the network and this is better achieved using a GPU unless wanting to wait for long hours. Simply launch the following command line (use the Docker image from https://cloud.docker.com/u/besn0847/repository/docker/besn0847/x86-tfcv2)

> python train_cnn.py -d dataset/ 

This will result into three files :

 - **cat_labels.bin** : the cat name label file
 - **cat_model_architecture.json** : the neural network architecture
 - **cat_neural_model.h5** : the neural network trained weights

A minimum 100+ pictures per animal is required as well as some thumbnails for false positives classified as 'nocat'.

### 3. Test your trained network
You can easily test your trained network with the following command line :

> python cnn.py -m cat_model_architecture.json -w cat_neural_model.h5 -l cat_labels.bin -i testset/bagguy/extract001.jpg

The result looks like :

> Cat: bagguy with probability: 0.9976345

We are now ready to proceed to expose the algorithm thru a webservice.

### 4. Expose the model thru a webservice
Simply use the startup script to launch a microservice  thru :

> docker run -ti --name catfaces -v ./Volumes/catfaces/:/conf/ -p 5000:5000 --entrypoint /startup.sh catfaces

And you can simply curl the image to get the same result :

> curl -X PUT -F image_file=@testset/bagguy/extract001.jpg
 http://localhost:5000/process

Result will be the same as in section 3.

 ### 5. Acknowledgment
 Most of the ideas and code is coming from [Adrian Rosebrock](https://twitter.com/PyImageSearch) at https://www.pyimagesearch.com/. I learnt a lot from your outstanding websites and tutorials. Merci ! 


CHANGES
****************************
*** May 2019 - Catson v1 ***
****************************
	. First implementation of Catson based on 96 x 96 ConvNets with 3 microservices
	. Trained on data collected in April and May 2019

****************************
*** August 2019 - Catson v2 ***
****************************
	. Second implementation of Catson based on 224 x 224 ConvNets with 3 microservices
	. Trained on data collected in July 2019 and tested on data collected in August 2019
	. Added 4 additional classes on top of 5 existing ones
	. Created Jupyter notebooks to ease dev / test / debug
	. Added PPT file to explain the overall concept
	. Did not change actual microservices (only on live system)


app.py
import cv2
import pickle
import argparse
import numpy as np
import tensorflow as tf
from PIL import Image
from keras.models import load_model
from keras.models import model_from_json
from keras.models import Sequential
from keras.layers.convolutional import Conv2D
from keras.layers.convolutional import MaxPooling2D
from keras.layers.core import Activation
from keras.layers.core import Flatten
from keras.layers.core import Dense
from keras.optimizers import Adam
from keras.preprocessing.image import img_to_array
from flask import Flask, render_template, request

app = Flask(__name__)
model = None

def init():
        print("Loading neural model")
        INPUT_MODEL_WEIGHTS = "/conf/cat_neural_model.h5"
        INPUT_MODEL_CONFIG = "/conf/cat_model_architecture.json" 
        INPUT_MODEL_CLASSES = "/conf/cat_labels.bin" 

        global graph
        graph = tf.get_default_graph() 
        #with open(INPUT_MODEL_CONFIG, 'r') as f:
        #       model = model_from_json(f.read())
        global model
        model = model_from_json(open(INPUT_MODEL_CONFIG, 'r').read())
        model.load_weights(INPUT_MODEL_WEIGHTS)

        global lb
        lb = pickle.loads(open(INPUT_MODEL_CLASSES, "rb").read())

@app.route('/process', methods=['POST','PUT'])      
def upload_file():                            
        file = request.files['image_file'] 
        image = cv2.imdecode(np.fromstring(file.read(), np.uint8), cv2.IMREAD_UNCHANGED)

        image = cv2.resize(image, (96, 96))
        image = image.astype("float") / 255.0
        image = img_to_array(image)
        image = np.expand_dims(image, axis=0)

        with graph.as_default():
            probas = model.predict(image)[0]

        idx = np.argmax(probas)
        label = lb.classes_[idx]
        proba = probas[idx]

        output = "Cat: " + label + " with probability:" + str(proba) + "\n"
        return output

if __name__ == '__main__':
        init()
        app.run(host='0.0.0.0')


extract.py
# import the necessary packages
from skimage.measure import compare_ssim
import argparse
import imutils
import cv2

ap = argparse.ArgumentParser()
ap.add_argument("-f", "--first", required=True,
	help="first input image")
ap.add_argument("-s", "--second", required=True,
	help="second")
args = vars(ap.parse_args())

# load the two input images
imageA = cv2.imread(args["first"])
imageB = cv2.imread(args["second"])
 
# convert the images to grayscale
grayA = cv2.cvtColor(imageA, cv2.COLOR_BGR2GRAY)
grayB = cv2.cvtColor(imageB, cv2.COLOR_BGR2GRAY)

# compute the Structural Similarity Index (SSIM) between the two
# images, ensuring that the difference image is returned
(score, diff) = compare_ssim(grayA, grayB, full=True)
diff = (diff * 255).astype("uint8")
#print("SSIM: {}".format(score))

# threshold the difference image, followed by finding contours to
# obtain the regions of the two input images that differ
thresh = cv2.threshold(diff, 0, 255,
	cv2.THRESH_BINARY_INV | cv2.THRESH_OTSU)[1]
cnts = cv2.findContours(thresh.copy(), cv2.RETR_EXTERNAL,
	cv2.CHAIN_APPROX_SIMPLE)
cnts = imutils.grab_contours(cnts)

# loop over the contours
for c in cnts:
	# compute the bounding box of the contour and then draw the
	# bounding box on both input images to represent where the two
	# images differ
	(x, y, w, h) = cv2.boundingRect(c)
	if w > 100 and h > 100:
#		cv2.rectangle(imageA, (x, y), (x + w, y + h), (0, 0, 255), 2)
		cv2.rectangle(imageB, (x, y), (x + w, y + h), (0, 0, 255), 2)
 
# show the output images
#cv2.imshow("Original", imageA)
cv2.imshow("Modified", imageB)
#cv2.imshow("Diff", diff)
#cv2.imshow("Thresh", thresh)
cv2.waitKey(0)


process.py
# import the necessary packages
from skimage.measure import compare_ssim
import argparse
import imutils
import cv2
from os import listdir
from os.path import isfile, join

ap = argparse.ArgumentParser()
ap.add_argument("-r", "--reference", required=True,
	help="reference image file")
ap.add_argument("-s", "--source", required=True,
	help="directory where to find all images to be processed")
ap.add_argument("-t", "--target", required=True,
	help="directory where to extract all deltas")
args = vars(ap.parse_args())

PREFIX = "extract"
SRCDIR = args["source"]
DSTDIR = args["target"]
REFERE = args["reference"]

IMGFILES = [f for f in listdir(SRCDIR) if isfile(join(SRCDIR, f))]

imageR = cv2.imread(REFERE)
count = 0

for f in IMGFILES:
	print("Processing:",join(SRCDIR, f))
	imageD = cv2.imread(join(SRCDIR, f))	

	grayA = cv2.cvtColor(imageR, cv2.COLOR_BGR2GRAY)
	grayB = cv2.cvtColor(imageD, cv2.COLOR_BGR2GRAY)

	(score, diff) = compare_ssim(grayA, grayB, full=True)
	diff = (diff * 255).astype("uint8")
	print("SSIM: {}".format(score))

	thresh = cv2.threshold(diff, 0, 255, cv2.THRESH_BINARY_INV | cv2.THRESH_OTSU)[1]
	cnts = cv2.findContours(thresh.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
	cnts = imutils.grab_contours(cnts)

	for c in cnts:
		(x, y, w, h) = cv2.boundingRect(c)
		if w > 100 and h > 100:
			count += 1
			#cv2.rectangle(imageD, (x, y), (x + w, y + h), (0, 0, 255), 2)
			print(join(DSTDIR,PREFIX+str(count)+".jpg"))
			cv2.imwrite(join(DSTDIR,PREFIX+str(count)+".jpg"), imageD[y:y+h, x:x+w])



train_cnn.py
# import the necessary packages
from keras.models import Sequential
from keras.layers.convolutional import Conv2D
from keras.layers.convolutional import MaxPooling2D
from keras.layers.core import Activation
from keras.layers.core import Flatten
from keras.layers.core import Dense
from keras.optimizers import Adam
from sklearn.preprocessing import LabelBinarizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from PIL import Image
from imutils import paths
import numpy as np
import argparse
import os
import pickle
 
# construct the argument parser and parse the arguments
ap = argparse.ArgumentParser()
ap.add_argument("-d", "--dataset", type=str, default="3scenes",
	help="path to directory containing the '3scenes' dataset")
args = vars(ap.parse_args())

# grab all image paths in the input dataset directory, then initialize
# our list of images and corresponding class labels
print("[INFO] loading images...")
imagePaths = paths.list_images(args["dataset"])
data = []
labels = []
 
# loop over our input images
for imagePath in imagePaths:
	# load the input image from disk, resize it to 32x32 pixels, scale
	# the pixel intensities to the range [0, 1], and then update our
	# images list
	image = Image.open(imagePath)
	image = np.array(image.resize((96, 96))) / 255.0
	data.append(image)
 
	# extract the class label from the file path and update the
	# labels list
	label = imagePath.split(os.path.sep)[-2]
	labels.append(label)

# encode the labels, converting them from strings to integers
lb = LabelBinarizer()
labels = lb.fit_transform(labels)
 
# perform a training and testing split, using 75% of the data for
# training and 25% for evaluation
(trainX, testX, trainY, testY) = train_test_split(np.array(data),
	np.array(labels), test_size=0.25)

# 20/01/2019 definition architecture reseau de neurones
model = Sequential()

model.add(Conv2D(8, (3, 3), padding="same", input_shape=(96, 96, 3)))
model.add(Activation("relu"))
model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))

model.add(Conv2D(16, (3, 3), padding="same"))
model.add(Activation("relu"))
model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))

model.add(Conv2D(32, (3, 3), padding="same"))
model.add(Activation("relu"))
model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))

model.add(Conv2D(64, (3, 3), padding="same"))
model.add(Activation("relu"))
model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))

model.add(Conv2D(128, (3, 3), padding="same"))
model.add(Activation("relu"))
model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))

model.add(Conv2D(256, (3, 3), padding="same"))
model.add(Activation("relu"))
model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))

model.add(Flatten())
model.add(Dense(5))
model.add(Activation("softmax"))
model.summary()

# train the model using the Adam optimizer
print("[INFO] training network...")
opt = Adam(lr=1e-3, decay=1e-3 / 50)
model.compile(loss="categorical_crossentropy", optimizer=opt,
	metrics=["accuracy"])
H = model.fit(trainX, trainY, validation_data=(testX, testY),
	epochs=50, batch_size=32)
 
# evaluate the network
print("[INFO] evaluating network...")
predictions = model.predict(testX, batch_size=32)
print(classification_report(testY.argmax(axis=1),
	predictions.argmax(axis=1), target_names=lb.classes_))

# saving neural model
model.save('cat_neural_model.h5')

with open('cat_model_architecture.json', 'w') as f:
    f.write(model.to_json())

f = open("cat_labels.bin", "wb")
f.write(pickle.dumps(lb))
f.close()


cnn.py
import cv2
import pickle
import argparse
import numpy as np
from PIL import Image
from keras.models import load_model
from keras.models import model_from_json
from keras.models import Sequential
from keras.layers.convolutional import Conv2D
from keras.layers.convolutional import MaxPooling2D
from keras.layers.core import Activation
from keras.layers.core import Flatten
from keras.layers.core import Dense
from keras.optimizers import Adam
from keras.preprocessing.image import img_to_array

ap = argparse.ArgumentParser()
ap.add_argument("-m", "--model", required=True,
	help="json file with trained model to use")
ap.add_argument("-w", "--weights", required=True,
	help="h5 file with trained model to use")
ap.add_argument("-l", "--labels", required=True,
	help="image labels")
ap.add_argument("-i", "--image", required=True,
	help="image to analyse")
args = vars(ap.parse_args())

with open(args["model"], 'r') as f:
	model = model_from_json(f.read())

model .load_weights(args["weights"])
image = cv2.imread(args["image"])
lb = pickle.loads(open(args["labels"], "rb").read())

image = cv2.resize(image, (96, 96))
image = image.astype("float") / 255.0
image = img_to_array(image)
image = np.expand_dims(image, axis=0)

probas = model.predict(image)[0]

idx = np.argmax(probas)
label = lb.classes_[idx]
proba = probas[idx]

print("Cat: " + label + " with probability:" + str(proba))
